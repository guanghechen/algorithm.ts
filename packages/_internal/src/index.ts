export * from './constant'
export * from './types'
