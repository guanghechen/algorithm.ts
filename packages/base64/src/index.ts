import { Base64 } from './base64'

export * from './base64'

const base64 = new Base64()

export const encode = base64.encode.bind(base64)
export const decode = base64.decode.bind(base64)
export const validate = base64.validate.bind(base64)
export default { encode, decode, validate }
