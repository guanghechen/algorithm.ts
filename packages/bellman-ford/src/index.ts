import { createCircularQueue } from '@algorithm.ts/circular-queue'

export interface IEdge {
  /**
   * The other end of the edge.
   */
  to: number
  /**
   * The cost of walking along this side.
   */
  cost: number
}

export interface IGraph {
  /**
   * The number of nodes in the graph. (0-index)
   */
  N: number
  /**
   * The source node. (0-index)
   */
  source: number
  /**
   * Graph edges.
   */
  edges: ReadonlyArray<IEdge>
  /**
   * Adjacency list. G[i] represent the index list of the  edges start from node i.
   */
  G: ReadonlyArray<ReadonlyArray<number>>
  /**
   * An array recording the shortest distance to the source point.
   */
  dist: number[]
}

export interface IOptions {
  /**
   * A big number, representing the unreachable cost.
   */
  INF?: number
  /**
   * Used to check if an element is already in the queue.
   */
  inq?: boolean[]
  /**
   * Record the number of times an element is enqueued,
   * used to check whether there is a negative cycle.
   */
  inqTimes?: number[]
}

const ZERO = 0
const DEFAULT_INF = Math.floor(Number.MAX_SAFE_INTEGER / 2)
const Q = createCircularQueue<number>()

/**
 * The bellman-ford algorithm, optimized with queue.
 *
 * @param graph
 * @param options
 */
export function bellmanFord(graph: IGraph, options: IOptions = {}): boolean {
  const { N, source, edges, G, dist } = graph
  const { INF = DEFAULT_INF, inq = [], inqTimes = [] } = options

  if (dist.length < N) dist.length = N
  if (inq.length < N) inq.length = N
  if (inqTimes.length < N) inqTimes.length = N

  dist.fill(INF, 0, N)
  inq.fill(false, 0, N)
  inqTimes.fill(0, 0, N)

  Q.init(N + 1)
  Q.enqueue(source)
  dist[source] = ZERO

  while (Q.size() > 0) {
    const o: number = Q.dequeue()!
    inq[o] = false

    for (const idx of G[o]) {
      const edge: IEdge = edges[idx]
      if (dist[edge.to] > dist[o] + edge.cost) {
        dist[edge.to] = dist[o] + edge.cost
        if (!inq[edge.to]) {
          Q.enqueue(edge.to)
          inq[edge.to] = true

          // eslint-disable-next-line no-plusplus
          if (++inqTimes[edge.to] > N) return false
        }
      }
    }
  }
  return true
}

export default bellmanFord
