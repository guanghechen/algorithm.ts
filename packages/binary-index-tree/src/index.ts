export * from './tree1'
export * from './tree2'
export * from './types'
export * from './util'
