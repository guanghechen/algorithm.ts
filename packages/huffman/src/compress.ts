export function compress(data: ReadonlyArray<0 | 1>): Uint8Array {
  const N: number = data.length
  const total = Math.ceil(N / 8)
  const result: Uint8Array = new Uint8Array(total + 1)

  // How many valid bits of the last data byte.
  result[0] = N & 7 ? N & 7 : 8

  let t = 0
  for (let i = 1; i < total; ++i, t += 8) {
    const value: number =
      (data[t] << 7) |
      (data[t + 1] << 6) |
      (data[t + 2] << 5) |
      (data[t + 3] << 4) |
      (data[t + 4] << 3) |
      (data[t + 5] << 2) |
      (data[t + 6] << 1) |
      data[t + 7]
    result[i] = value
  }

  let value = 0
  for (; t < data.length; ++t) value = (value << 1) | data[t]
  result[total] = value
  return result
}

export function decompress(data: Uint8Array): Array<0 | 1> {
  const size: number = data.length
  const result: Array<0 | 1> = []
  for (let i = 1, _end = size - 1; i < _end; ++i) {
    const v: number = data[i]
    for (let x = 0x80; x > 0; x >>= 1) result.push(v & x ? 1 : 0)
  }

  // Last data byte.
  {
    const validBits: number = data[0]
    if (validBits > 0) {
      const v: number = data[size - 1]
      for (let x = 1 << (validBits - 1); x > 0; x >>= 1) result.push(v & x ? 1 : 0)
    }
  }
  return result
}
