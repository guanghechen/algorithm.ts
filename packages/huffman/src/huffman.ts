import { createPriorityQueue } from '@algorithm.ts/priority-queue'

export interface IHuffmanNode {
  value?: string
  left?: IHuffmanNode
  right?: IHuffmanNode
}

export type IHuffmanEncodedData = Array<0 | 1>

export type IHuffmanEncodingTable = Record<string, IHuffmanEncodedData>

/**
 * Create a Huffman tree for the specified texts.
 */
export function createHuffmanTree(text: string): IHuffmanNode {
  const costMap: Record<string, number> = {}
  for (const c of text) {
    const cnt: number = costMap[c] ?? 0
    costMap[c] = cnt + 1
  }

  const entries = Object.entries(costMap)
  if (entries.length <= 0) return {}

  const priorityQueue = createPriorityQueue<{
    cost: number
    node: IHuffmanNode
  }>((x, y) => y.cost - x.cost)
  priorityQueue.init(entries.map(([value, cost]) => ({ cost, node: { value } })))
  while (priorityQueue.size() > 1) {
    const o1 = priorityQueue.dequeue()!
    const o2 = priorityQueue.dequeue()!
    const o: IHuffmanNode = { left: o1.node, right: o2.node }
    priorityQueue.enqueue({
      cost: o1.cost + o2.cost,
      node: o,
    })
  }
  return priorityQueue.top()!.node
}

/**
 * Build a HuffmanEncodingTable based on Huffman tree.
 * @param tree
 * @returns
 */
export function buildEncodingTable(tree: IHuffmanNode): IHuffmanEncodingTable {
  const prefix: IHuffmanEncodedData = []
  const encodingTable: IHuffmanEncodingTable = {}
  collect(tree, 0)
  return encodingTable

  function collect(node: IHuffmanNode, cur: number): void {
    if (node.value) encodingTable[node.value] = prefix.slice(0, cur)
    else {
      if (node.left) {
        prefix[cur] = 0
        collect(node.left, cur + 1)
      }
      if (node.right) {
        prefix[cur] = 1
        collect(node.right, cur + 1)
      }
    }
  }
}

/**
 * Build a huffman tree based on HuffmanEncodingTable.
 * @param table
 * @returns
 */
export function buildHuffmanTree(table: IHuffmanEncodingTable): IHuffmanNode {
  const entries = Object.entries(table)
  if (entries.length === 0) return {}

  const root: IHuffmanNode = {}
  for (const [value, path] of entries) {
    let o = root
    for (const x of path) {
      if (x === 0) {
        if (o.left === undefined) o.left = {}
        o = o.left
      } else {
        if (o.right === undefined) o.right = {}
        o = o.right
      }
    }
    o.value = value
  }
  return root
}
