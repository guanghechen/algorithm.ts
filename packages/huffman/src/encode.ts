import { buildEncodingTable, createHuffmanTree } from './huffman'
import type { IHuffmanEncodedData, IHuffmanEncodingTable, IHuffmanNode } from './huffman'

/**
 * Encode texts through Huffman encode algorithm.
 * @param plaintext
 * @returns
 */
export function encode(plaintext: string): {
  encodedData: IHuffmanEncodedData
  encodingTable: IHuffmanEncodingTable
  tree: IHuffmanNode
} {
  const tree: IHuffmanNode = createHuffmanTree(plaintext)
  const encodedData: Array<0 | 1> = []
  const encodingTable: IHuffmanEncodingTable = buildEncodingTable(tree)
  for (const c of plaintext) {
    // Invariant: data should not be null / undefined.
    const data: IHuffmanEncodedData = encodingTable[c]
    encodedData.push(...data)
  }
  return { encodedData, encodingTable, tree }
}

/**
 * @param encodedData
 * @param tree
 * @returns
 */
export function decode(encodedData: IHuffmanEncodedData, tree: IHuffmanNode): string {
  const L: number = encodedData.length
  if (L <= 0) return ''

  let i = 0
  let o: IHuffmanNode | undefined
  let plainText = ''

  while (i < L) {
    for (o = tree; o && o.value === undefined; ++i) {
      o = encodedData[i] === 0 ? o.left! : o.right!
    }
    if (o?.value === undefined) break
    plainText += o.value
  }

  if (i !== L || o?.value === undefined) {
    throw new TypeError('Bad encoded data or huffman tree.')
  }
  return plainText
}
