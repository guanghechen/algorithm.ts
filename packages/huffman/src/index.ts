import { compress, decompress } from './compress'
import { decode, encode } from './encode'
import { buildEncodingTable, buildHuffmanTree, createHuffmanTree } from './huffman'

export * from './compress'
export * from './encode'
export * from './huffman'

export default {
  encode,
  decode,
  compress,
  decompress,
  buildEncodingTable,
  createHuffmanTree,
  buildHuffmanTree,
}
