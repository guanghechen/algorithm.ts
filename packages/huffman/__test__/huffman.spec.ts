import { compress, decode, decompress, encode, fromEncodingTable } from '../src'

describe('basic', function () {
  it('empty', () => textWrapper(''))
  it('Hello, world!', () => textWrapper('Hello, world!'))
  it('Hello, world!1', () => textWrapper('Hello, world!1'))
  it('Hello, world!11', () => textWrapper('Hello, world!11'))
  it('Hello, world!111', () => textWrapper('Hello, world!111'))

  it('中文', () => textWrapper('中文'))

  function textWrapper(text: string): void {
    const encodeResult = encode(text)
    expect(encodeResult).toMatchSnapshot()

    const { encodedData, encodingTable, tree } = encodeResult

    const text2: string = decode(encodedData, tree)
    expect(text2).toEqual(text)

    const tree2 = fromEncodingTable(encodingTable)
    expect(tree).toEqual(tree2)

    expect(decode(encodedData, tree2)).toEqual(text)

    const compressedData = compress(encodedData)
    expect(compressedData).toMatchSnapshot()

    expect(decompress(compressedData)).toEqual(encodedData)
  }
})

it('unexpected', function () {
  const { encodedData, encodingTable } = encode('Hello, world!')
  const { H, ...encodingTable2 } = encodingTable
  const tree2 = fromEncodingTable(encodingTable2)

  expect(() => decode(encodedData, tree2)).toThrow(/Bad encoded data or huffman tree/)
})
