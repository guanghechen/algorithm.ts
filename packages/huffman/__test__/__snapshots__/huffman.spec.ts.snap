// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic Hello, world! 1`] = `
Object {
  "encodedData": Array [
    1,
    1,
    1,
    0,
    1,
    0,
    0,
    1,
    0,
    1,
    0,
    1,
    1,
    1,
    0,
    0,
    0,
    1,
    1,
    0,
    1,
    1,
    1,
    0,
    1,
    0,
    1,
    1,
    0,
    0,
    0,
    0,
    0,
    1,
    1,
    0,
    0,
    0,
    1,
    1,
    1,
    1,
  ],
  "encodingTable": Object {
    " ": Array [
      1,
      0,
      1,
      1,
    ],
    "!": Array [
      1,
      1,
      1,
      1,
    ],
    ",": Array [
      0,
      0,
      1,
    ],
    "H": Array [
      1,
      1,
      1,
      0,
    ],
    "d": Array [
      1,
      0,
      0,
      0,
    ],
    "e": Array [
      1,
      0,
      0,
      1,
    ],
    "l": Array [
      0,
      1,
    ],
    "o": Array [
      1,
      1,
      0,
    ],
    "r": Array [
      0,
      0,
      0,
    ],
    "w": Array [
      1,
      0,
      1,
      0,
    ],
  },
  "tree": Object {
    "left": Object {
      "left": Object {
        "left": Object {
          "value": "r",
        },
        "right": Object {
          "value": ",",
        },
      },
      "right": Object {
        "value": "l",
      },
    },
    "right": Object {
      "left": Object {
        "left": Object {
          "left": Object {
            "value": "d",
          },
          "right": Object {
            "value": "e",
          },
        },
        "right": Object {
          "left": Object {
            "value": "w",
          },
          "right": Object {
            "value": " ",
          },
        },
      },
      "right": Object {
        "left": Object {
          "value": "o",
        },
        "right": Object {
          "left": Object {
            "value": "H",
          },
          "right": Object {
            "value": "!",
          },
        },
      },
    },
  },
}
`;

exports[`basic Hello, world! 2`] = `
Uint8Array [
  2,
  233,
  92,
  110,
  176,
  99,
  3,
]
`;

exports[`basic Hello, world!1 1`] = `
Object {
  "encodedData": Array [
    1,
    0,
    0,
    0,
    0,
    1,
    1,
    0,
    0,
    0,
    0,
    0,
    1,
    1,
    0,
    0,
    1,
    1,
    1,
    0,
    1,
    0,
    1,
    0,
    0,
    1,
    1,
    1,
    0,
    1,
    1,
    1,
    0,
    0,
    0,
    1,
    1,
    1,
    1,
    1,
    0,
    1,
    1,
    1,
    0,
    1,
    0,
  ],
  "encodingTable": Object {
    " ": Array [
      0,
      1,
      0,
    ],
    "!": Array [
      1,
      0,
      1,
      1,
    ],
    ",": Array [
      0,
      1,
      1,
      1,
    ],
    "1": Array [
      1,
      0,
      1,
      0,
    ],
    "H": Array [
      1,
      0,
      0,
      0,
    ],
    "d": Array [
      1,
      1,
      1,
      1,
    ],
    "e": Array [
      0,
      1,
      1,
      0,
    ],
    "l": Array [
      0,
      0,
    ],
    "o": Array [
      1,
      1,
      0,
    ],
    "r": Array [
      1,
      1,
      1,
      0,
    ],
    "w": Array [
      1,
      0,
      0,
      1,
    ],
  },
  "tree": Object {
    "left": Object {
      "left": Object {
        "value": "l",
      },
      "right": Object {
        "left": Object {
          "value": " ",
        },
        "right": Object {
          "left": Object {
            "value": "e",
          },
          "right": Object {
            "value": ",",
          },
        },
      },
    },
    "right": Object {
      "left": Object {
        "left": Object {
          "left": Object {
            "value": "H",
          },
          "right": Object {
            "value": "w",
          },
        },
        "right": Object {
          "left": Object {
            "value": "1",
          },
          "right": Object {
            "value": "!",
          },
        },
      },
      "right": Object {
        "left": Object {
          "value": "o",
        },
        "right": Object {
          "left": Object {
            "value": "r",
          },
          "right": Object {
            "value": "d",
          },
        },
      },
    },
  },
}
`;

exports[`basic Hello, world!1 2`] = `
Uint8Array [
  7,
  134,
  12,
  234,
  119,
  31,
  58,
]
`;

exports[`basic Hello, world!11 1`] = `
Object {
  "encodedData": Array [
    0,
    1,
    0,
    0,
    0,
    1,
    1,
    1,
    0,
    0,
    0,
    0,
    1,
    0,
    1,
    1,
    1,
    1,
    0,
    1,
    1,
    1,
    1,
    1,
    0,
    0,
    0,
    1,
    0,
    1,
    1,
    0,
    0,
    1,
    0,
    0,
    0,
    1,
    1,
    0,
    0,
    1,
    0,
    1,
    1,
    1,
    0,
    1,
    1,
    0,
  ],
  "encodingTable": Object {
    " ": Array [
      1,
      1,
      1,
      1,
    ],
    "!": Array [
      0,
      1,
      0,
      1,
    ],
    ",": Array [
      1,
      1,
      1,
      0,
    ],
    "1": Array [
      1,
      1,
      0,
    ],
    "H": Array [
      0,
      1,
      0,
      0,
    ],
    "d": Array [
      0,
      1,
      1,
      0,
    ],
    "e": Array [
      0,
      1,
      1,
      1,
    ],
    "l": Array [
      0,
      0,
    ],
    "o": Array [
      1,
      0,
      1,
    ],
    "r": Array [
      1,
      0,
      0,
      1,
    ],
    "w": Array [
      1,
      0,
      0,
      0,
    ],
  },
  "tree": Object {
    "left": Object {
      "left": Object {
        "value": "l",
      },
      "right": Object {
        "left": Object {
          "left": Object {
            "value": "H",
          },
          "right": Object {
            "value": "!",
          },
        },
        "right": Object {
          "left": Object {
            "value": "d",
          },
          "right": Object {
            "value": "e",
          },
        },
      },
    },
    "right": Object {
      "left": Object {
        "left": Object {
          "left": Object {
            "value": "w",
          },
          "right": Object {
            "value": "r",
          },
        },
        "right": Object {
          "value": "o",
        },
      },
      "right": Object {
        "left": Object {
          "value": "1",
        },
        "right": Object {
          "left": Object {
            "value": ",",
          },
          "right": Object {
            "value": " ",
          },
        },
      },
    },
  },
}
`;

exports[`basic Hello, world!11 2`] = `
Uint8Array [
  2,
  71,
  11,
  223,
  22,
  70,
  93,
  2,
]
`;

exports[`basic Hello, world!111 1`] = `
Object {
  "encodedData": Array [
    0,
    1,
    1,
    0,
    1,
    1,
    0,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    1,
    0,
    0,
    1,
    0,
    1,
    0,
    1,
    0,
    1,
    1,
    0,
    1,
    0,
    0,
    1,
    0,
    0,
    0,
    1,
    0,
    1,
    1,
    1,
    1,
    1,
    1,
    0,
    0,
    0,
    1,
    1,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "encodingTable": Object {
    " ": Array [
      1,
      0,
      1,
      1,
    ],
    "!": Array [
      0,
      1,
      1,
      1,
    ],
    ",": Array [
      1,
      0,
      1,
      0,
    ],
    "1": Array [
      0,
      0,
    ],
    "H": Array [
      0,
      1,
      1,
      0,
    ],
    "d": Array [
      1,
      1,
      0,
      0,
    ],
    "e": Array [
      1,
      1,
      0,
      1,
    ],
    "l": Array [
      1,
      1,
      1,
    ],
    "o": Array [
      1,
      0,
      0,
    ],
    "r": Array [
      0,
      1,
      0,
      1,
    ],
    "w": Array [
      0,
      1,
      0,
      0,
    ],
  },
  "tree": Object {
    "left": Object {
      "left": Object {
        "value": "1",
      },
      "right": Object {
        "left": Object {
          "left": Object {
            "value": "w",
          },
          "right": Object {
            "value": "r",
          },
        },
        "right": Object {
          "left": Object {
            "value": "H",
          },
          "right": Object {
            "value": "!",
          },
        },
      },
    },
    "right": Object {
      "left": Object {
        "left": Object {
          "value": "o",
        },
        "right": Object {
          "left": Object {
            "value": ",",
          },
          "right": Object {
            "value": " ",
          },
        },
      },
      "right": Object {
        "left": Object {
          "left": Object {
            "value": "d",
          },
          "right": Object {
            "value": "e",
          },
        },
        "right": Object {
          "value": "l",
        },
      },
    },
  },
}
`;

exports[`basic Hello, world!111 2`] = `
Uint8Array [
  5,
  109,
  254,
  85,
  164,
  95,
  142,
  0,
]
`;

exports[`basic empty 1`] = `
Object {
  "encodedData": Array [],
  "encodingTable": Object {},
  "tree": Object {},
}
`;

exports[`basic empty 2`] = `
Uint8Array [
  0,
]
`;

exports[`basic 中文 1`] = `
Object {
  "encodedData": Array [
    0,
    1,
  ],
  "encodingTable": Object {
    "中": Array [
      0,
    ],
    "文": Array [
      1,
    ],
  },
  "tree": Object {
    "left": Object {
      "value": "中",
    },
    "right": Object {
      "value": "文",
    },
  },
}
`;

exports[`basic 中文 2`] = `
Uint8Array [
  2,
  1,
]
`;
